src/transformers/models/vits/modeling_vits.py

```diff
@@ -1487,3 +1766,420 @@ def forward(
+
+            [What are attention masks?](../glossary#attention-mask)
+        monotic_alignment_function (`Callable`, *optional*):
+            Monotonic alignment function. Used for training, i.e when `labels` are provided. By default, it will use a
+            Pytorch implementation of the monotonic alignment function which is awfully slow. An alternative relying on
+            cython is proposed in examples/pytorch/text-to-speech/run_vits_finetuning.py
+        Returns:
+
+        Example:
+
+        ```python
+        >>> from transformers import VitsTokenizer, VitsModelForPreTraining, set_seed
+        >>> import torch
+
+        >>> tokenizer = VitsTokenizer.from_pretrained("facebook/mms-tts-eng")
+        >>> model = VitsModelForPreTraining.from_pretrained("facebook/mms-tts-eng")
+
+        >>> inputs = tokenizer(text="Hello - my dog is cute", return_tensors="pt")
+
+        >>> set_seed(555)  # make deterministic
+
+        >>> with torch.no_grad():
+        ...     outputs = model(inputs["input_ids"])
+        >>> outputs.waveform.shape
+        torch.Size([1, 45824])
+        ```
+        """
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        monotic_alignment_function = (
+            monotonic_align_max_path if monotic_alignment_function is None else monotic_alignment_function
+        )
+
+        if attention_mask is not None:
+            input_padding_mask = attention_mask.unsqueeze(-1).float()
+        else:
+            input_padding_mask = torch.ones_like(input_ids).unsqueeze(-1).float()
+
+        if self.config.num_speakers > 1 and speaker_id is not None:
+            if isinstance(speaker_id, int):
+                speaker_id = torch.full(size=(1,), fill_value=speaker_id, device=self.device)
+            elif isinstance(speaker_id, (list, tuple, np.ndarray)):
+                speaker_id = torch.tensor(speaker_id, device=self.device)
+
+            if not ((0 <= speaker_id).all() and (speaker_id < self.config.num_speakers).all()).item():
+                raise ValueError(f"Set `speaker_id` in the range 0-{self.config.num_speakers - 1}.")
+            if not (len(speaker_id) == 1 or len(speaker_id == len(input_ids))):
+                raise ValueError(
+                    f"You passed {len(speaker_id)} `speaker_id` but you should either pass one speaker id or `batch_size` `speaker_id`."
+                )
+
+            speaker_embeddings = self.embed_speaker(speaker_id).unsqueeze(-1)
+        else:
+            speaker_embeddings = None
+
+        # if inference, return inference forward of VitsModel
+        if labels is None:
+            return self._inference_forward(
```

Comment:
Is it necessary to have a whole separate method here? It seems there's a lot of overlap in the logic. It also breaks the common `forward` pattern in other models and makes the logic overall harder to maintain as it has to be updated in two places

Reply:
Well, for a few reasons:
1. the original repo has a separate method, I simply replicated the logic
2. there are IMO a lot of difference
3. it was easier to debug !

I could indeed fuse the two methods (to train and to infer) but IMO, it would be even more difficult to maintain - here we only have to replicate part of the `VitsModel.forward` on `_inference_forward`. If I fused the two methods, it would be more difficult for sure.

I'll listen to your suggestions of course! What do you think would be the best option?
1. Fuse training and inference logic
2. Only have the training logic in `VitsModelForPretraining` and leave the inference code for `VitsModel`
3. Leave it as it is

-------

I'll also put the inference and train diagram here just in case for reference:

**TRAINING**

![image](https://github.com/huggingface/transformers/assets/52246514/c6f02870-ef21-49e0-9e6e-3e46585a9eb9)

**INFERENCE**

![image](https://github.com/huggingface/transformers/assets/52246514/8d32319d-82a0-41d6-b634-5b5066dfde41)

Reply:
OK, I see, thanks for sharing the diagram! 

In general, we shouldn't just be following the design patterns of other repos. Part of the advantage of transformers is that it provides a standardised interface and code structure - so sometimes we have to re-work quite a bit to keep this. 

That being said, if there's large differences, we don't want to make the forward pass too hard to understand. I'm happy to leave as-is 👍